In [105]:
# importing the necessary Library
import random
import pandas as pd
import nltk
from nltk.corpus import movie_reviews

In [106]:
#Importing the DataSet
review = [(list(movie_reviews.words(fileid)),category)
            for category in movie_reviews.categories()
            for fileid in movie_reviews.fileids(category)]

In [107]:
# Viewing the Data
rint('Total Number of Reviews: {}'.format(len(review)))
print('The Fisrt Review: {}'.format(review[0]))

Total Number of Reviews: 2000
The Fisrt Review: ([u'plot', u':', u'two', u'teen', u'couples', u'go', u'to', u'a', u'church', u'party', u',', u'drink', u'and', u'then', u'drive', u'.', u'they', u'get', u'into', u'an', u'accident', u'.', u'one', u'of', u'the', u'guys', u'dies', u',', u'but', u'his', u'girlfriend', u'continues', u'to', u'see', u'him', u'in', u'her', u'life', u',', u'and', u'has', u'nightmares', u'.', u'what', u"'", u's', u'the', u'deal', u'?', u'watch', u'the', u'movie', u'and', u'"', u'sorta', u'"', u'find', u'out', u'.', u'.', u'.', u'critique', u':', u'a', u'mind', u'-', u'fuck', u'movie', u'for', u'the', u'teen', u'generation', u'that', u'touches', u'on', u'a', u'very', u'cool', u'idea', u',', u'but', u'presents', u'it', u'in', u'a', u'very', u'bad', u'package', u'.', u'which', u'is', u'what', u'makes', u'this', u'review', u'an', u'even', u'harder', u'one', u'to', u'write', u',', u'since', u'i', u'generally', u'applaud', u'films', u'which', u'attempt', u'to', u'break'

In [110]:
random.shuffle(review)

In [111]:
# Creating a Frequency map for the words in the DataSet
all_words=[]
for word in movie_reviews.words():
    all_words.append(word)
all_words = nltk.FreqDist(all_words)

print('Total Number of Unique Words: {}'.format(len(all_words)))
print('Top 15 Most Common Words: {}'.format(all_words.most_common(15)))

Total Number of Unique Words: 39768
Top 15 Most Common Words: [(u',', 77717), (u'the', 76529), (u'.', 65876), (u'a', 38106), (u'and', 35576), (u'of', 34123), (u'to', 31937), (u"'", 30585), (u'is', 25195), (u'in', 21822), (u's', 18513), (u'"', 17612), (u'it', 16107), (u'that', 15924), (u'-', 15595)]


**We need to build our feature set form the given words. But it will be impractical to use the whole set of 39768 words to build features. So we are going to use the 2500 most common words from the given words to train our model.**

In [112]:
features = list(all_words.keys())[:2500]
print(features[:5])

[u'sonja', u'askew', u'woods', u'spiders', u'bazooms']


In [120]:
# defining a function to create feature_set
def find_features(doc):
    words = set(doc)
    dic = {}
    for word in words:
        dic[word] = (word in features)
    return dic

In [121]:
feature_set = [(find_features(rev),category) for (rev,category) in review]

In [122]:
# spiliting the data in Training and Testing
from sklearn.model_selection import train_test_split

training,testing = train_test_split(feature_set,random_state=97)

In [123]:
len(training),len(testing)

(1500, 500)

In [124]:
#importing the necessary classifiers
from nltk.classify.scikitlearn import SklearnClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

In [126]:
model = SklearnClassifier(LogisticRegression())
model.train(training)

C:\Users\HP\Anaconda2\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


<SklearnClassifier(LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False))>

In [128]:
accuracy = nltk.classify.accuracy(model,testing)
print('Accuracy Score for Logistic Regression: {}'.format(accuracy))

Accuracy Score for Logistic Regression: 0.69


In [129]:
model = SklearnClassifier(KNeighborsClassifier(n_neighbors=5))
model.train(training)
accuracy = nltk.classify.accuracy(model,testing)
print('Accuracy Score for K Nearest Neighbour: {}'.format(accuracy))

Accuracy Score for K Nearest Neighbour: 0.582


In [130]:
model = SklearnClassifier(DecisionTreeClassifier(max_depth=5,max_leaf_nodes=10))
model.train(training)
accuracy = nltk.classify.accuracy(model,testing)
print('Accuracy Score for Decision Tree: {}'.format(accuracy))

Accuracy Score for Decision Tree: 0.592


In [131]:
model = SklearnClassifier(SVC(kernel='linear'))
model.train(training)
accuracy = nltk.classify.accuracy(model,testing)
print('Accuracy Score for Support Vector Machine: {}'.format(accuracy))

Accuracy Score for Support Vector Machine: 0.64


**From the above observations we conclude that Logistic Regressor has the best Accuracy**